File: code/generate_strategies/Direct/cpp/function_Cerberus_Validator_schema.cpp
#include <string>
#include <algorithm>
#include <nlohmann/json.hpp>
#include <type_traits>

class DataValidator {
public:
    using json = nlohmann::json;
    
    
    template<typename T>
    std::string validate_data(const T& data_input) {
        if constexpr (std::is_same_v<T, std::string>) {
            // 将输入中的单引号替换为双引号，以适应 JSON 解析要求
            std::string fixed = data_input;
            std::replace(fixed.begin(), fixed.end(), '\'', '\"');
            
            json data;
            try {
                data = json::parse(fixed);
            }
            catch (const json::parse_error& e) {
                return "Error: Invalid data - {\"parse_error\": [\"" + std::string(e.what()) + "\"]}";
            }
            
            // 校验字段 name
            if (data.contains("name")) {
                if (!data["name"].is_string()) {
                    return "Error: Invalid data - {'name': ['must be of string type']}";
                } else {
                    std::string name = data["name"].get<std::string>();
                    if (name.length() < 3) {
                        return "Error: Invalid data - {'name': ['min length is 3']}";
                    }
                }
            }
            // 校验字段 age
            if (data.contains("age")) {
                if (!data["age"].is_number_integer()) {
                    return "Error: Invalid data - {'age': ['must be of integer type']}";
                } else {
                    int age = data["age"].get<int>();
                    if (age < 18) {
                        return "Error: Invalid data - {'age': ['min value is 18']}";
                    }
                    else if (age > 100) {
                        return "Error: Invalid data - {'age': ['max value is 100']}";
                    }
                }
            }
            
            return "Valid data: {'name': {'type': 'string', 'minlength': 3}, 'age': {'type': 'integer', 'min': 18, 'max': 100}}";
        } else {
            return "Error: Input must be a string";
        }
    }
};

File: manual/code/generate_strategies/IR(pseudocode)/java/java/function_Cerberus_Validator_schema.java
package com.example;

import com.fasterxml.jackson.databind.ObjectMapper;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Arrays;

public class DataValidator {

    public String validate_data(String dataStr) {
        
        class DataValidator {
            private Map<String, String[]> validateData(Map<String, Object> data) {
                Map<String, String[]> errors = new LinkedHashMap<>();

                
                Object nameValue = data.get("name");
                if (nameValue == null) {
                    errors.put("name", new String[]{"must be of string type"});
                } else if (!(nameValue instanceof String)) {
                    errors.put("name", new String[]{"must be of string type"});
                } else if (((String) nameValue).length() < 3) {
                    errors.put("name", new String[]{"min length is 3"});
                }

                
                Object ageValue = data.get("age");
                if (ageValue == null) {
                    errors.put("age", new String[]{"must be of integer type"});
                } else {
                    int age;
                    if (ageValue instanceof Integer) {
                        age = (Integer) ageValue;
                    } else if (ageValue instanceof Number) {
                        age = ((Number) ageValue).intValue();
                    } else {
                        errors.put("age", new String[]{"must be of integer type"});
                        return errors;
                    }

                    if (age < 18) {
                        errors.put("age", new String[]{"min value is 18"});
                    } else if (age > 100) {
                        errors.put("age", new String[]{"max value is 100"});
                    }
                }

                return errors;
            }
        }

        try {
            
            Map<String, Map<String, Object>> schema = new LinkedHashMap<>();
            Map<String, Object> nameRules = new LinkedHashMap<>();
            nameRules.put("type", "string");
            nameRules.put("minlength", 3);

            Map<String, Object> ageRules = new LinkedHashMap<>();
            ageRules.put("type", "integer");
            ageRules.put("min", 18);
            ageRules.put("max", 100);

            schema.put("name", nameRules);
            schema.put("age", ageRules);

            
            ObjectMapper mapper = new ObjectMapper();
            Map<String, Object> data = mapper.readValue(dataStr.replace("'", "\""), Map.class);

            
            DataValidator validator = new DataValidator();
            Map<String, String[]> errors = validator.validateData(data);

            if (!errors.isEmpty()) {
                
                StringBuilder errorMsg = new StringBuilder();
                errorMsg.append("Error: Invalid data - {");
                boolean first = true;
                for (Map.Entry<String, String[]> entry : errors.entrySet()) {
                    if (!first) {
                        errorMsg.append(", ");
                    }
                    errorMsg.append(entry.getKey())
                            .append("=[")
                            .append(String.join(", ", entry.getValue()))
                            .append("]");
                    first = false;
                }
                errorMsg.append("}");
                return errorMsg.toString();
            }

            return "Valid data: " + schema.toString();
        } catch (Exception e) {
            if (e.getMessage() != null && e.getMessage().contains("Cannot deserialize value of type `java.lang.Integer` from String")) {
                return "Error: Invalid data - {age=[must be of integer type]}";
            }
            return "Error: Invalid data - {" + e.getMessage() + "}";
        }
    }
}